### PR TITLE
Spanize interop in System.Net.NetworkInformation

### PR DESCRIPTION
**Pull request by jkotas**
Faster and safer

### CODE REVIEW

**Comment by jkotas** [comment]
Depends on #35078

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/SystemTcpConnection.cs [hunk 1]
```diff
@@ -11,7 +11,7 @@ internal class SystemTcpConnectionInformation : TcpConnectionInformation
         private readonly IPEndPoint _remoteEndPoint;
         private readonly TcpState _state;
 
-        internal SystemTcpConnectionInformation(Interop.IpHlpApi.MibTcpRow row)
+        internal SystemTcpConnectionInformation(in Interop.IpHlpApi.MibTcpRow row)
```
Comment (jkotas) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/SystemTcpConnection.cs:14]: We will initialize the connection information directly from the buffer that was filled in by the OS API, without multiple intermediate copies and allocations.

**Review by wfurt — APPROVED** [review]
LGTM

**Comment by Wraith2** [comment]
Which bits make it faster and which safer? It looks like it's adding unsafe modifiers (for the fixed) which makes is counter intuitive to be safer from a casual perspective.

**Review by stephentoub — APPROVED** [review]
Nice.

**Comment by stephentoub** [comment]
> Which bits make it faster and which safer? It looks like it's adding unsafe modifiers (for the fixed) which makes is counter intuitive to be safer from a casual perspective.

The faster part is e.g. not using array marshaling and PtrToStructure, and using sizeof instead of Marshal.SizeOf.

The safer part is using span slicing instead of math on pointers, with the spans being bounds checked.

**Comment by jkotas** [comment]
> unsafe modifiers (for the fixed) which makes is counter intuitive to be safe

`unsafe` modifier is required for unsafe C# language constructs, but not for unsafe APIs. It makes a poor indicator for whether the code is unsafe or now. #31354 should make it better.

**Pull request merged**